File: scripts/Math/Matrix23.lua
require "Object"
---@class Matrix23:Object
Matrix23 = {
    __parent = Object
}

setmetatable(Matrix23, {
    __call = function(self, e00, e01, e02, e10, e11, e12)
        ---@class Matrix23
        local obj = {}
        setmetatable(obj, {
            __index = self,
            __mul = self.__mul
        })
        obj:_construct(e00, e01, e02, e10, e11, e12)
        return obj
    end,
    __index = Matrix23.__parent
})

function Matrix23:_construct(e00, e01, e02, e10, e11, e12)
    Matrix23.__parent._construct(self)
    e00 = e00 or 0
    e01 = e01 or 0
    e02 = e02 or 0
    e10 = e10 or 0
    e11 = e11 or 0
    e12 = e12 or 0
    self.m00, self.m10 = e00, e10
    self.m01, self.m11 = e01, e11
    self.m02, self.m12 = e02, e12
end

function Matrix23:setTranslation(a)
    self.m00, self.m10 = 1, 0
    self.m01, self.m11 = 0, 1
    self.m02, self.m12 = a.x, a.y
end

-- 从 x 转 y, 与 cpp 不同, 我觉得是 cpp 取了
function Matrix23:setRotation(r)
    r = r * math.pi / 180
    local c = math.cos(r)
    local s = math.sin(r)
    self.m00, self.m10 = c, s
    self.m01, self.m11 = -s, c
    self.m02, self.m12 = 0, 0
end

function Matrix23:setScaling(a)
    self.m00, self.m10 = a.x, 0
    self.m01, self.m11 = 0, a.y
    self.m02, self.m12 = 0, 0
end

-- ok
function Matrix23:scale(a)
    self.m00, self.m10 = self.m00 * a.x, self.m10 * a.y
    self.m01, self.m11 = self.m01 * a.x, self.m11 * a.y
    self.m02, self.m12 = self.m02 * a.x, self.m12 * a.y
end

-- ok
function Matrix23:translate(a)
    self.m02 = self.m02 + a.x
    self.m12 = self.m12 + a.y
end

-- ok
function Matrix23:rotate(r)
    local c = math.cos(r * math.pi / 180);
    local s = math.sin(r * math.pi / 180);
    self.m00, self.m10 = self.m00 * c - self.m10 * s, self.m00 * s + self.m10 * c;
    self.m01, self.m11 = self.m01 * c - self.m11 * s, self.m01 * s + self.m11 * c;
    self.m02, self.m12 = self.m02 * c - self.m12 * s, self.m02 * s + self.m12 * c;
end

function Matrix23:__mul(m)
    local m00, m10 = self.m00 * m.m00 + self.m10 * m.m01, self.m00 * m.m10 + self.m10 * m.m11
    local m01, m11 = self.m01 * m.m00 + self.m11 * m.m01, self.m01 * m.m10 + self.m11 * m.m11
    local m02, m12 = self.m02 * m.m00 + self.m12 * m.m01 + m.m02, self.m02 * m.m10 + self.m12 * m.m11 + m.m12
    self.m00, self.m10 = m00, m10
    self.m01, self.m11 = m01, m11
    self.m02, self.m12 = m02, m12
    return self
end

-- 没有什么问题
function Matrix23:multiply(out, input)
    local tx = input.x
    out.x = self.m00 * tx + self.m01 * input.y + self.m02
    out.y = self.m10 * tx + self.m11 * input.y + self.m12
end

File: Example/Matrix23.cpp
#include "Matrix23.h"
#include "Vector2.h"
#include "GameLib/Math.h"
using namespace GameLib;

Matrix23::Matrix23(){
}

void Matrix23::setTranslation( const Vector2& a ){
	m00 = m11 = 1.0;
	m01 = m10 = 0.0;
	m02 = a.x;
	m12 = a.y;
}

void Matrix23::setRotation( double r ){
	double s = GameLib::sin( r );
	double c = GameLib::cos( r );
	m00 = m11 = c;
	m01 = s;
	m10 = -s;
	m02 = m12 = 0.0;
}

void Matrix23::setScaling( const Vector2& a ){
	m00 = a.x;
	m11 = a.y;
	m01 = m10 = m02 = m12 = 0.0;
}

void Matrix23::operator*=( const Matrix23& m ){
	double tx, ty;
	tx = m00; ty = m01;
	m00 = tx * m.m00 + ty * m.m10;
	m01 = tx * m.m01 + ty * m.m11;
	m02 = tx * m.m02 + ty * m.m12 + m02;
	tx = m10; ty = m11;
	m10 = tx * m.m00 + ty * m.m10;
	m11 = tx * m.m01 + ty * m.m11;
	m12 = tx * m.m02 + ty * m.m12 + m12;
}

void Matrix23::multiply( Vector2* out, const Vector2& in ) const {
	double tx = in.x; //out和in可能相同，因此备份
	out->x = m00 * tx + m01 * in.y + m02; //左上方* x +中间上方* y +右上方* 1
	out->y = m10 * tx + m11 * in.y + m12; //左下* x +中间下* y +左下* 1
}

//和移动矩阵相乘的做法

void Matrix23::translate( const Vector2& a ){
	m02 += m00 * a.x + m01 * a.y; 
	m12 += m10 * a.x + m11 * a.y;
}

//和缩放矩阵相乘的做法
/*
a b c
d e f

に

sx 0 0
0 sy 0

を掛ける。

a b c   sx 0  0
d e f   0  sy 0
0 0 1   0  0  1

の上2行を計算すれば、

a*sx b*sy c
d*sx e*sy f

これが答えである。
m00,m01,m10,m11の4要素を触ればいい
*/
void Matrix23::scale( const Vector2& a ){
	m00 *= a.x;
	m01 *= a.y;
	m10 *= a.x;
	m11 *= a.y;
}

//和旋转旋转矩阵相乘的做法
/*
a b c
d e f

に

p -q 0
q p 0

を掛ける。pがコサインで、qがサインだ。

a b c   p -q 0
d e f   q  p 0
0 0 1   0  0 1

上2行を計算すると、

(ap + bq) (-aq + bp) c
(dp + eq) (-dq + ep) f

m00,m01,m10,m11の4要素を触ればいい
*/

void Matrix23::rotate( double r ){
	double c = GameLib::cos( r );
	double s = GameLib::sin( r );
	//m00需要备份。因为第一行覆盖了m00。
	double t = m00;
	m00 = t * c + m01 * s;
	m01 = t * ( -s ) + m01 * c;
	t = m10; //m10备份
	m10 = t * c + m11 * s;
	m11 = t * ( -s ) + m11 * c;
}
